### PR TITLE
GH-725: Fix return Collection<Message<?>>

### PR DESCRIPTION
**Pull request by garyrussell**
Fixes https://github.com/spring-projects/spring-kafka/issues/725
Fixes https://github.com/spring-projects/spring-kafka/issues/726

Handle return type `Collection<Message<?>>`.
Allow empty `@SendTo` when the return type is `Message<?>` (or a collection thereof).

### CODE REVIEW

File: src/reference/asciidoc/kafka.adoc [hunk 1]
```diff
@@ -1171,6 +1171,22 @@ public KafkaTemplate<String, String> myReplyingTemplate() {
 }
 ----
 
+IMPORTANT: If the listener method returns `Message<?>` or `Collection<Messge<?>>` the listener method is responsible for setting up the message headers for the reply; for example, when handling a request from a `ReplyingKafkaTemplate`, you might do the following:
```
Comment (artembilan) [inline src/reference/asciidoc/kafka.adoc]: When `if`, `when` etc is in the first part of the sentence, it must be separated with a comma : https://www.ego4u.com/en/cram-up/writing/comma?10 😉 

I mean in this case we need comma before `the listener method`.

File: src/reference/asciidoc/kafka.adoc [hunk 2]
```diff
@@ -1171,6 +1171,22 @@ public KafkaTemplate<String, String> myReplyingTemplate() {
...
+
+[source, java]
+----
+@KafkaListener(id = "messageReturned", topics = "someTopic")
+public Message<?> listen(String in, @Header(KafkaHeaders.REPLY_TOPIC) byte[] replyTo,
```
Comment (artembilan) [inline src/reference/asciidoc/kafka.adoc]: Are you missing `@SendTo` here?
Or it doesn't matter and non-`void` return is enough to produce reply?

Comment (garyrussell) [reply src/reference/asciidoc/kafka.adoc]: Interesting; yes; I guess if the return type is `Message<?>` the `@SendTo` is entirely optional - we could enforce it, I suppose, but maybe it's ok. 

WDYT?

File: src/reference/asciidoc/kafka.adoc (hunk 1 above)
Comment (garyrussell) [reply src/reference/asciidoc/kafka.adoc]: Hehe - thanks, yes, please fix on merge (unless you want me to do something with your other comment).

File: src/reference/asciidoc/kafka.adoc (hunk 2 above)
Comment (artembilan) [reply src/reference/asciidoc/kafka.adoc]: Ah! Yes, it's really OK: your sentence really implies that.
So, yeah, merging...
Thanks

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-kafka/commit/d9c23e3c32e2687b9022ce6d1f72ce2955ecce0a

**Pull request closed without merging**